File: android-compose/app/src/main/java/com/motchill/androidcompose/data/repository/DefaultMotchillRepository.kt
package com.motchill.androidcompose.data.repository

import android.util.Log
import com.motchill.androidcompose.core.network.MotchillApiClient
import com.motchill.androidcompose.core.security.MotchillPlayCipher
import com.motchill.androidcompose.domain.model.HomeSection
import com.motchill.androidcompose.domain.model.MovieDetail
import com.motchill.androidcompose.domain.model.NavbarItem
import com.motchill.androidcompose.domain.model.PlaySource
import com.motchill.androidcompose.domain.model.PopupAdConfig
import com.motchill.androidcompose.domain.model.SearchFilterData
import com.motchill.androidcompose.domain.model.SearchResults

class DefaultMotchillRepository(
    private val apiClient: MotchillApiClient,
) : MotchillRepository {
    override suspend fun loadHome(): List<HomeSection> = apiClient.fetchHomeSections()

    override suspend fun loadNavbar(): List<NavbarItem> = apiClient.fetchNavbar()

    override suspend fun loadDetail(slug: String): MovieDetail = apiClient.fetchMovieDetail(slug)

    override suspend fun loadPreview(slug: String): MovieDetail = apiClient.fetchMoviePreview(slug)

    override suspend fun loadSearchFilters(): SearchFilterData = apiClient.fetchSearchFilters()

    override suspend fun loadSearchResults(
        categoryId: Int?,
        countryId: Int?,
        typeRaw: String,
        year: String,
        orderBy: String,
        isChieuRap: Boolean,
        is4k: Boolean,
        search: String,
        pageNumber: Int,
    ): SearchResults {
        return apiClient.fetchSearchResults(
            categoryId = categoryId,
            countryId = countryId,
            typeRaw = typeRaw,
            year = year,
            orderBy = orderBy,
            isChieuRap = isChieuRap,
            is4k = is4k,
            search = search,
            pageNumber = pageNumber,
        )
    }

    override suspend fun loadEpisodeSources(
        movieId: Int,
        episodeId: Int,
        server: Int,
    ): List<PlaySource> {
        val payload = apiClient.fetchEpisodeSourcesPayload(
            movieId = movieId,
            episodeId = episodeId,
            server = server,
        )
        Log.d(
            TAG,
            buildString {
                append("loadEpisodeSources movieId=")
                append(movieId)
                append(" episodeId=")
                append(episodeId)
                append(" server=")
                append(server)
                append(" payloadLength=")
                append(payload.length)
                append(" payloadPreview=")
                append(payload.take(80))
            },
        )
        val sources = MotchillPlayCipher.decodeSources(payload)
        Log.d(
            TAG,
            buildString {
                append("decodedSources count=")
                append(sources.size)
                append(" items=")
                append(
                    sources.joinToString(separator = " | ") { source ->
                        buildString {
                            append("id=")
                            append(source.sourceId)
                            append(",frame=")
                            append(source.isFrame)
                            append(",quality=")
                            append(source.quality)
                            append(",server=")
                            append(source.serverName)
                            append(",subtitleField=")
                            append(source.subtitle)
                            append(",tracks=")
                            append(source.tracks.size)
                            append(",audioTracks=")
                            append(source.audioTracks.size)
                            append(",subtitleTracks=")
                            append(source.subtitleTracks.size)
                            append(",defaultAudio=")
                            append(source.defaultAudioTrack?.displayLabel.orEmpty())
                            append(",defaultSubtitle=")
                            append(source.defaultSubtitleTrack?.displayLabel.orEmpty())
                            if (source.tracks.isNotEmpty()) {
                                append(",trackDetails=[")
                                append(
                                    source.tracks.joinToString(separator = "; ") { track ->
                                        buildString {
                                            append("kind=")
                                            append(track.kind)
                                            append(",label=")
                                            append(track.label)
                                            append(",file=")
                                            append(track.file)
                                            append(",default=")
                                            append(track.isDefault)
                                            append(",isAudio=")
                                            append(track.isAudio)
                                            append(",isSubtitle=")
                                            append(track.isSubtitle)
                                        }
                                    },
                                )
                                append("]")
                            }
                        }
                    },
                )
            },
        )
        return sources
    }

    override suspend fun loadPopupAd(): PopupAdConfig? = apiClient.fetchPopupAd()

    companion object {
        private const val TAG = "Motchill.player"
    }
}

File: swiftUI/Data/Repositories/DefaultMotchillRepository.swift
import Foundation

final class DefaultMotchillRepository: MotchillRepository, @unchecked Sendable {
    private let apiClient: MotchillAPIClient

    init(apiClient: MotchillAPIClient) {
        self.apiClient = apiClient
    }

    func loadHome() async throws -> [MotchillHomeSection] {
        try await apiClient.fetchHomeSections()
    }

    func loadNavbar() async throws -> [MotchillNavbarItem] {
        try await apiClient.fetchNavbar()
    }

    func loadDetail(slug: String) async throws -> MotchillMovieDetail {
        try await apiClient.fetchMovieDetail(slug: slug)
    }

    func loadPreview(slug: String) async throws -> MotchillMovieDetail {
        try await apiClient.fetchMoviePreview(slug: slug)
    }

    func loadSearchFilters() async throws -> MotchillSearchFilterData {
        try await apiClient.fetchSearchFilters()
    }

    func loadSearchResults(
        categoryId: Int?,
        countryId: Int?,
        typeRaw: String,
        year: String,
        orderBy: String,
        isChieuRap: Bool,
        is4k: Bool,
        search: String,
        pageNumber: Int
    ) async throws -> MotchillSearchResults {
        try await apiClient.fetchSearchResults(
            categoryId: categoryId,
            countryId: countryId,
            typeRaw: typeRaw,
            year: year,
            orderBy: orderBy,
            isChieuRap: isChieuRap,
            is4k: is4k,
            search: search,
            pageNumber: pageNumber
        )
    }

    func loadEpisodeSources(
        movieID: Int,
        episodeID: Int,
        server: Int
    ) async throws -> [MotchillPlaySource] {
        try await apiClient.fetchEpisodeSources(
            movieId: movieID,
            episodeId: episodeID,
            server: server
        )
    }

    func loadPopupAd() async throws -> MotchillPopupAdConfig? {
        try await apiClient.fetchPopupAd()
    }
}
